lasagne/utils.py

```diff
@@ -337,105 +355,91 @@ def create_param(spec, shape, name=None):
         # `name` attribute of the variable, but the user may have already
         # named the variable and we don't want to override this.
         if spec.ndim != len(shape):
-            raise RuntimeError("parameter variable has %d dimensions, "
-                               "should be %d" % (spec.ndim, len(shape)))
+            raise ValueError("%s has %d dimensions, should be %d" %
+                             (err_prefix % "Theano variable", spec.ndim,
+                              len(shape)))
+        if not spec.name:
+            spec.name = name
         return spec
 
-    elif isinstance(spec, np.ndarray):
-        if spec.shape != shape:
-            raise RuntimeError("parameter array has shape %s, should be "
-                               "%s" % (spec.shape, shape))
-        return theano.shared(spec, name=name)
-
-    elif hasattr(spec, '__call__'):
-        arr = spec(shape)
-        try:
-            arr = floatX(arr)
-        except Exception:
-            raise RuntimeError("cannot initialize parameters: the "
-                               "provided callable did not return an "
-                               "array-like value")
-        if arr.shape != shape:
-            raise RuntimeError("cannot initialize parameters: the "
-                               "provided callable did not return a value "
-                               "with the correct shape")
-        return theano.shared(arr, name=name)
-
     else:
-        raise RuntimeError("cannot initialize parameters: 'spec' is not "
-                           "a numpy array, a Theano expression, or a "
-                           "callable")
+        if hasattr(spec, '__call__'):
```

Comment:
This is not going to work, `spec` has been replaced by then. As in my proposal, it should be `if "callable" in err_prefix`. It's a bit ugly, but otherwise the error message is less precise.